vscode-dotty/src/tracer.ts

```diff
@@ -0,0 +1,370 @@
+        if (this.isTracingEnabled && this.tracingConsent.get() === 'no-answer') this.askForTracingConsent()
+        this.initializeAsyncWorkspaceDump()
+        const lspOutputChannel = this.createLspOutputChannel()
+        const statusBarItem = this.createStatusBarItem()
+        for (const disposable of [consentCommandDisposable, lspOutputChannel, statusBarItem]) {
+            if (disposable) this.ctx.extensionContext.subscriptions.push(disposable)
+        }
+        return { lspOutputChannel }
+    }
+
+    private askForTracingConsent(): void {
+        vscode.window.showInformationMessage(
+            'Do you want to help EPFL develop Dotty LSP plugin by uploading your LSP communication? ' +
+            'PLEASE BE AWARE that the data sent contains your entire codebase and ALL the IDE actions, ' +
+            'including every single keystroke.',
+            'yes', 'no'
+        ).then((value: string | undefined) => {
+            if (value === 'yes' || value === 'no') this.tracingConsent.set(value)
+        })
+    }
+
+    private initializeAsyncWorkspaceDump() {
+        if (this.remoteWorkspaceDumpUrl === undefined) return
+        // convince TS that this is a string
+        const definedUrl: string = this.remoteWorkspaceDumpUrl
+
+        const doInitialize = () => {
+            try {
+                this.asyncUploadWorkspaceDump(definedUrl)
+            } catch (err) {
+                this.logError('error during workspace dump', safeError(err))
+            }
+        }
+
+        if (this.tracingConsent.get() === 'yes') {
+            doInitialize()
+        } else {
+            let didInitialize = false
+            this.tracingConsent.subscribe(() => {
+                if (didInitialize) return
+                didInitialize = true
+                doInitialize()
+            })
+        }
+    }
+
+    private createStatusBarItem(): vscode.StatusBarItem | undefined {
+        if (!this.isTracingEnabled) return undefined
+        const item = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Left, 0)
+        item.command = consentCommandName
+        const renderStatusBarItem = () => {
+            item.text = (() => {
+                const desc = this.tracingConsent.get() === 'yes' ? 'ON' : 'OFF'
+                return `$(radio-tower) Dotty trace: ${desc}`
+            })()
+
+            item.tooltip = (() => {
+                const desc = this.tracingConsent.get() === 'yes' ? 'consented' : 'not consented'
+                return `This workspace is configured for remote tracing of Dotty LSP and you have ${desc} to it. ` +
+                    'Click to adjust your consent.'
+            })()
+        }
+        renderStatusBarItem()
+        this.tracingConsent.subscribe(renderStatusBarItem)
+        item.show()
+        return item
+    }
+
+    private createLspOutputChannel(): vscode.OutputChannel | undefined {
+        if (!this.remoteTracingUrl) return undefined
```

Comment:
Unless I'm missing something, it looks like this means that we don't get any outputchannel, even a local one, when remoteTracingUrl is undefined. Suggestion: make an LspOutputChannel that extends OutputChannel and take as argument a (local) OutputChannel, then every method of LspOutputChannel just forwards to the underlying OutputChannel and also does network stuff if tracing is enabled.